### DISCUSSION TITLE
Is client routing 'middleware' folder currently supported anymore? #Nuxt3

### DISCUSSION THREAD

**Discussion by hannanmiah** [nuxt/nuxt#16767] [Questions] [answered]
What is the current solution of client side middleware with Nuxt 3? What is the current status of router middleware support to Nuxt 3? I can't find any topics on Nuxt webpage and anywhere related to this.

**Comment by Norel** [comment, marked as answer]
I'm using this in the layout script to redirect to another page (login page for example).
```
<script setup>
    const router = useRouter()

    router.beforeEach((to, from) => {
        const conditionIsMet = false
        if (to.path === '/original/path' && !conditionIsMet) {
            return '/redirect/to/path'
        }
    })
</script>
```

**Reply by wobsoriano** [reply]
Yes, this is how it should be done now.

**Reply by hannanmiah** [reply]
But, it would have been better if there have any support out of the box. I still have to do it manually. By the way, thanks for this solution.

**Reply by AXSJ** [reply]
I tried this but it calls beforeEach to add a function to the router beforeEach on every page load, so the beforeEach calls keep piling up... what am I doing wrong?

**Reply by majutaca** [reply]
I tried this it seems the beforeEach only runs if someone clicks the link to the route, however it does not run when a user goes directly to the URL using the browser address bar. This means that if a user knows the URL they will be able to access the page.

**Reply by danielroe** [reply]
This discussion is outdated. You can add a global middleware. See docs: https://v3.nuxtjs.org/docs/directory-structure/middleware/

**Reply by steffenstolze** [reply]
@danielroe I'm still facing an issue with a simple middleware. 

I have two layouts, `auth` (for everything auth-related) and `default` (for all other routes).

I have the following client middleware in place

```typescript
export default defineNuxtRouteMiddleware((to, _from) => {
  const user = useSupabaseUser();

  // Redirect anonymous user to /auth/signin if user tries to access anything but auth pages
  if (!user.value && !to.path.includes('/auth/')) {
    return navigateTo('/auth/signin');
  }

  // /auth/signin and /auth/signup not reachable for already logged in users
  if (user.value && (to.path === '/auth/signin' || to.path === '/auth/signup')) {
    return navigateTo('/');
  }
});
```

When I access the page at root, e.g. `http://localhost:3000`, the client middleware does its job and redirects to `/auth/signin` **but** not without a short preview of the `/` route, displaying the default layout.
The slower the connection, the longer the `/` route with default layout is displayed, before navigating to `/auth/signin`

I don't want users to ever see the default layout and associated routes when they are not logged in, not even for half a second.

Tbh, I don't understand why the default layout is loaded at all, since I assumed that the middleware would catch it before entering the route.

The app is configured as SPA

```javascript
export default defineNuxtConfig({
  ssr: false,
  modules: ['@nuxtjs/tailwindcss', '@nuxtjs/supabase'],
  build: {
    transpile: ['@heroicons/vue'],
  },
});
```

thanks!

**Comment by eMeRiKa13** [comment]
Thank you for your answer, I managed to make it work.

I had to change `if (to === '/original/path' && !conditionIsMet)` to `if (to.path === '/original/path' && !conditionIsMet)`

**Comment by eMeRiKa13** [comment]
When I navigate in my app it's working well, but if I hit directly a specific page in my browser like http://localhost:3000/dashboard/home the function ` router.beforeEach` doesn't seem to be called. 

Is-it normal? how to manage that?

**Reply by exreplay** [reply]
If you also want to make it work server side you could also use the `response` from the `ssrContext` which you can get by calling `useNuxtApp`. This way you are able to get the current url, resolve it with vue-router and redirect with the server response. Something like this:

```ts
const { ssrContext } = useNuxtApp();
const router = useRouter();
const conditionIsMet = false;

if (ssrContext) {
  const { res, url } = ssrContext;
  const to = router.resolve(url);

  if (to === '/original/path' && !conditionIsMet) {
    res.writeHead(302, {
      Location: '/redirect/to/path'
    });
    res.end();
  }
}

router.beforeEach((to, from) => {
  if (to === '/original/path' && !conditionIsMet) {
    return '/redirect/to/path';
  }
});
```

This is what i am currently using and it works fine.

**Reply by eMeRiKa13** [reply]
Awesome, I made it work. (I had to use `to.path ===` and not only `to ===`)

Thanks a lot

**Reply by exreplay** [reply]
You are welcome, glad i could help 👌

**Reply by maxdzin** [reply]
Greetings, guys!
A very nice example. I've tried to use it this way also, and If I use router.beforeEach on pages - it works well.
But if I'm trying to use it in layout script setup it is caused pages to refresh when I'm trying to navigate between the pages. It looks like the router is not working anymore in this case but I'm not yet managed to debug that. And also, there are no errors in the console in this case.
Do you recognize the same behavior and if so, have you managed to overcome it?

**Reply by maxdzin** [reply]
@exreplay , @eMeRiKa13

**Reply by eMeRiKa13** [reply]
I use it only in layout script setup, and it works

**Reply by maxdzin** [reply]
@eMeRiKa13 I've double-checked and updated to the latest dependencies and it works now.
Thank you guys a lot!
